### ISSUE TITLE
[Enhancement] Provide images for ARM64

### ISSUE THREAD

**Issue by alexandru-tetelea** [strimzi/strimzi-kafka-operator#3357]
Hi guys, I'm sure that it can't be your main priority, but could you provide images for arm64 also? Thank you

**Comment by scholzj** [comment]
I have it on my todo list. Hope to find some time for it later this summer (but no promises - you know how it is :-o). Also, right now we do not have any environments to develop and test on arm64. Also I think we have zero experience with arm64.

**Comment by scholzj** [comment]
I think the enhancement title is right here ;-)

**Comment by alexandru-tetelea** [comment]
Thank you for your response Jakub. I will be happy to test it on my Raspberry pi 4 with microk8s
Agree Enhancement is better :)

**Comment by bitvijays** [comment]
@scholzj Hope you are doing well. Thank you for the awesome support. Is there any updates on this?.

We would be happy to help to resolve this over Christmas break, if could be guided a bit? We need to run Apache Kafka/ ZooKeeper, MirrorMaker on Raspberry Pi 4 and Jetson Nano and require a automated deployment.

For example: How to start?, Probably we need changes at [docker images](https://github.com/strimzi/strimzi-kafka-operator/tree/master/docker-images) ? Should we start with reading [Development Docs](https://github.com/strimzi/strimzi-kafka-operator/tree/master/development-docs)

**Comment by scholzj** [comment]
I do not know what your environment is. You can now build Strimzi images with `docker buildx`: https://github.com/strimzi/strimzi-kafka-operator/blob/master/development-docs/DEV_GUIDE.md#building-container-images-for-other-platforms-with-docker-buildx ... that should allow you to build the Arm images directly on Intel PCs which makes things easier.

To get it working on ARM, you still need to replace some of the binaries directly downloaded with the AArch64 versions. The the `tini` binary and the Kafka exporter binary are the two which come to my mind (they are both available, you just need to change the URL and the checksum).

I did some tests in the past -> after doing this you can quite easily get it running on Arm64. But I did not do any detailed tests. Kafka has lot of binary dependencies in its JARs (all the different compressions etc.) so it could be that some of them might not work if they do not contain Arm binaries. That would be something which would need to fixed with the dependencies directly.

I hope to find some time for this soon as well. My target is the same ... get it running on RPi4 or in Arm based clouds. The next step I wanted to look at was to make it easier to deal with the binaries above.

For the record, I do not think we have currently any test environments for Arm. So until that is solved, I do not think we plan to have any official Strimzi Arm images. But I want to at least make it easy to build and maybe have some unofficial repos with the Arm images.

**Comment by nderraugh** [comment]
@bitvijays the basic process is outlined here https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/.  

Similar recent(ish) work for prometheus-operator can be found here https://github.com/prometheus-operator/prometheus-operator/pull/3177.

**Comment by bitvijays** [comment]
Thanks @nderraugh and @scholzj :) Happy New Year :)

@scholzj Trying to compile `strimzi` from source, got the error mentioned in https://github.com/strimzi/strimzi-kafka-operator/issues/4167 Will need a bit of guidance.

**Comment by bitvijays** [comment]
@scholzj  Hope you are doing well. Thank you for your guidance. 

I understand that `arm64` image is currently not supported and this also because of unavailability of the test environments for Arm. 

On the other hand, we are aware that the changes are required in the Docker images of `jmxtrans` , `kafka`, `operator` for `tini` binary and the `Kafka exporter` binary. Would it be possible to change the `DockerFile` and enable CI/CD for creation of `amd64` as well as `arm64` images. Further, officially, we can say that: `arm64` images of `strimzi` has not been tested and are not guaranteed to work.

Also, I would be happy to help to test the images by pulling from `strimzi:arm64` tag. This could be one-step closer to `arm64`. I am happy to help in this. 

Yours Sincerely,
bitvijays

**Comment by scholzj** [comment]
As I suggested in the past - I think we should have a way how to parametrize the platform in the build process, to make it possible to build the Arm64 images without changing the Dockerfiles. That should make it easier for you to build your own images for Arm and use them. I think we should do that by adding some parametrization to the URLs and checksums of the respective artifacts (tini, kafka-exporter are the affected ones). I might have a look into it once I get to it. But if anyone wants to work on it and contribute it before I get to it, you are of course free to do so. But as I said before - I do not think we should push any untested images into our container repositories.

**Comment by bitvijays** [comment]
@scholzj 

Would that be something like the below? based on `TARGETPLATFORM`? If so, I can give it a try.
```
RUN set -ex; \
    if [[ ${TARGETPLATFORM} = "linux/arm64" ]]; then \
        wget -O /usr/bin/tini https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-arm64; \
        echo "7c5463f55393985ee22357d976758aaaecd08defb3c5294d353732018169b019 */usr/bin/tini" | sha256sum -c; \
        chmod +x /usr/bin/tini; \
    else \
        wget -O /usr/bin/tini https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini; \
        echo "12d20136605531b09a2c2dac02ccee85e1b874eb322ef6baf7561cd93f93c855 */usr/bin/tini" | sha256sum -c; \
        chmod +x /usr/bin/tini; \
    fi
```

**Comment by scholzj** [comment]
For example. I was maybe originally thinking about having multiple vars with the platform as a suffix or something. But this looks fine as well. We can also inject them from outside possibly. But you got the idea right => to be able to pull the right binary automatically without changing anything.

**Comment by XBeg9** [comment]
I am happy to be beta tester for arm64 images

**Comment by derketo** [comment]
I am tentatively (insufficient experience) trying to port a minikube/aks application which uses kafka via strimzi operator and images to microk8s on a raspi cluster. I have microk8s doing its thing, but of course now need strimzi images for arm64.

I notice that at least one of the strimzi dockerfiles in github now has ARCHITECTURE config for arm64 as discussed above in this thread. 

Please can someone quickly describe the current state of this process/subproject/whatever ?

As per XBeg9, also happy to do some beta testing, although my lack of knowledge may limit helpfulness of bug reports or insights. Always good to learn and I offer another willing pair of hands who may catch further issues if my use case is sufficiently different to other volunteers.

Thanks.

**Comment by scholzj** [comment]
@derketo You can try to follow this guide for building Arm64 images: https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/DEV_GUIDE.md#building-container-images-for-other-platforms-with-docker-buildx

**Comment by scholzj** [comment]
This has been done in #5991 and the main branch has now support for Arm64 / AArch64. It will be released as part of the 0.27.0 release. We call the Arm64 support currently as _experimental_ since we need to build the CI to have a full test coverage. If you want, you can help us test them: The CI build now produces multi-arch images, so you should now need anything special, just use the installation YAMLs from `packaging/install/cluster-operator` in the main branch and check that it works in your environments and for your use-cases.
